federationapi/routing/send.go

```diff
@@ -211,8 +216,27 @@ func (t *txnReq) processEventWithMissingState(e gomatrixserverlib.Event) error {
 		return err
 	}
 	// Check that the event is allowed by the state.
+retryAllowedState:
 	if err := checkAllowedByState(e, state.StateEvents); err != nil {
-		return err
+		switch missing := err.(type) {
+		case gomatrixserverlib.MissingAuthEventError:
+			// An auth event was missing so let's look up that event over federation
+			for _, s := range state.StateEvents {
+				if s.EventID() == missing.AuthEventID {
+					if serr := t.processEventWithMissingState(s, depth+1); serr == nil {
+						// If there was no error retrieving the event from federation then
+						// we assume that it succeeded, so retry the original state check
```

Comment:
Should this retry count as an attempt? I feel like it probably should, or else this could potentially infinite loop.

Reply:
The outcomes are either:

1. We fail `t.federation.LookupState` at which point we give up because that won't return a `MissingAuthEventError`
1. We fail `state.Check` at which point we give up for the same reason as above
1. We pass both of the above and hit `checkAllowedByState` again, but this will only recurse if another event is actually missing (since if we retrieved the original one, it'll now satisfy or hard-fail the above two conditions)
1. If all of the above succeeds then no more recursions are made anyway

Reply:
I just don't like infinite loops which lack clear and obvious terminating cases, *especially* when the data is being provided by potentially malicious parties outside our control.

I can't give you a failure case, but I worry some pathological circular event refs could cause this code to run forever.

Reply:
Or alternatively even a case where the HS returns an empty set (without triggering an error) could cause this code to loop.

Reply:
I know what you mean, although this is apparently what Synapse does and I'm hasten to do something different in case we end up with obviously broken rooms as a result. 

Since the only event that can have neither prev events nor auth events are `m.room.create` events, anything else with empty sets should fail `state.Check` before reaching this point.

Reply:
Okay, I'll have a think about how we can best protect ourselves from this and other things that we need to do for Synapse parity but are just yucky.